Source/ocdm/adapter/rdk/open_cdm_adapter.cpp

```diff
@@ -0,0 +1,508 @@
+/*
+ * If not stated otherwise in this file or this component's LICENSE file the
+ * following copyright and licenses apply:
+ *
+ * Copyright 2020 RDK Management
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "open_cdm_adapter.h"
+#include "open_cdm_impl.h"
+
+#include "Module.h"
+#include <gst/gst.h>
+#include <gst/base/gstbytereader.h>
+
+#include <gst_svp_meta.h>
+#include "../CapsParser.h"
+
+//void Log(const char* file, int line, LogLevel ll, const char* fmt, ...);
+//void LogTelemetry(const char* file, int line,  int x);
+// Log APIs
+//#define LOGT(X) LogTelemetry(__FILE__, __LINE__, X)
+//#define LOGE(...) Log(__FILE__, __LINE__, CDMi::ERROR,  __VA_ARGS__)
+
+//#define LOGW(...) Log(__FILE__, __LINE__, CDMi::WARNING,  __VA_ARGS__)
+//#define LOGI(...) Log(__FILE__, __LINE__, CDMi::INFO, __VA_ARGS__)
+//#define LOGD(...) Log(__FILE__, __LINE__, CDMi::DEBUG,  __VA_ARGS__)
+//#define LOGV(...) Log(__FILE__, __LINE__, CDMi::VERBOSE,  __VA_ARGS__)
+
+typedef uint32_t SEC_SIZE;
+struct My_Sec_OpaqueBufferHandle_struct
+{
+    SEC_SIZE dataBufSize;
+    void *sess;
+    SEC_SIZE align;
+    int ion_fd;
+    int map_fd;
+    unsigned int register_shm;
+    void *rtkmem_handle;
+};
+
+EXTERNAL OpenCDMError opencdm_gstreamer_transform_caps(GstCaps** caps)
+{
+    OpenCDMError result (ERROR_NONE);
+
+    if(!gst_svp_ext_transform_caps(caps, TRUE))
+        result = ERROR_UNKNOWN;
+
+    return (result);
+}
+
+bool swapIVBytes(uint8_t *mappedIV,uint32_t mappedIVSize)
+{
+    uint8_t buf;
+    for (uint32_t i = 0; i < mappedIVSize / 2; i++) {
+        uint8_t buf = mappedIV[i];
+        mappedIV[i] = mappedIV[mappedIVSize - i - 1];
+        mappedIV[mappedIVSize - i - 1] = buf;      
+    }
+
+    return true;
+}
+
+uint32_t opencdm_construct_session_private(struct OpenCDMSession* session, void* &pvtData)
+{
+    bool success = gst_svp_ext_get_context(&pvtData, Server, (unsigned int)session);
+    char buf[25] = { 0 };
+    snprintf(buf, 25, "%X", (unsigned int)session);
+    ASSERT(session != nullptr);
+    session->SetParameter("rpcId", buf);
```

Comment:
Is this SetParameter still needed. I remember from our discussions that the session id string can be used as a unique identifier between the client side and the server side for these purposes.

Reply:
This is required for realtek platforms

Reply:
Would like to get this clarified.

Reply:
@sramani-metro @pwielders 
I have modified the rdk-adapter with following changes to take session if as a unique identifier for socket path creation. https://github.com/rdkcentral/ThunderClientLibraries/pull/250/commits/ce8d1981491ecbe57451801da39a55e56035ce2b 
But In client side, If we try to create a context in mediakeysession constructor, the client is trying to establish connection before the server side is ready which is causing the race condition. 
` CWVMediaKeySession::CWVMediaKeySession 
{
......
#ifdef USE_SVP
  if (!m_sessionId.empty()) {
    LOGW("Initializing SVP context for client side ID = %s\n", m_sessionId.c_str());
    gst_svp_ext_get_context(&m_pSVPContext, Client, m_sessionId.c_str());
  }
......
}`

Similarly we tried the same in first decrypt call which is causing issue due to race condition. 
` CDMi_RESULT CWVMediaKeySession::Decrypt(
        uint8_t*                 inData,          // Incoming encrypted data
        const uint32_t           inDataLength,    // Incoming encrypted data length
        uint8_t**                outData,         // Outgoing decrypted data
        uint32_t*                outDataLength,   // Outgoing decrypted data length
        const SampleInfo*        sampleInfo,      // Information required to decrypt Sample
        const IStreamProperties* properties)
{
#ifdef USE_SVP
  if (!m_svpContextInit && !m_sessionId.empty()) {
    LOGW("Initializing SVP context for client side ID = %s\n", m_sessionId.c_str());
    if (gst_svp_ext_get_context(&m_pSVPContext, Client, m_sessionId.c_str()))
	{
		m_svpContextInit = true;
	}

  }
  if (!m_pSVPContext)
  {
    LOGW("Failed to initialize SVP context\n");
  }
#endif
........
}`

Due to this we have tried to invoke the context creation in client side once it is done in server side using setParameter() API. Since setParameter() is already part of ThunderInterfaces Can we use the API for the connection establishment between server and client side. Please share your input regarding this

Reply:
Hi @kkanag314, @sramani-metro and myself discussed it. 
First an observation we like to share, how can it be that a less privileged process (the client, e.g. Netflix, that includes the OCDM client library, is a server, the client now needs the right to be able to create server sockets. I think a very undesirable situation, I would have left this up to the OCDM module, which is a more privileged process. So architecturally I think there is something terrible wrong here.
Having said that, suppose we need to cope with this situation, there is, we think, still a way forward. First of all the setParameter is not "already" part of the ThunderInterface, it is this pull request that would bring it in and we are in discussion if it is really needed. We prefer to keep the interface as slim as possible and hence why we do not see the actual use of this method (on the interface) yet.
What we (or ate least what @sramani-metro explained me) is that after the session has been created (OCDM server size), the User private data for the session is attached to it (client side), during this attaching of the User private data client side, the server for Realtek, on client side is started. The socket, with the SessionId is now open, the OCDM server side session now needs to hook up to the server, running client side.
Our remark here would be, why not, after you started the realtek server, call Load() on the Session interface, this call is forwarded to the OCDMServer, implemented in the DRM module side and thus, it would be perfect timing to than connect, in this implementation to the socket (SessionId).
I learned that @sramani-metro reported in ill, so get well soon, but if he is up and running again, he can definitely help you out with the details.
The Load call on OCDMclient library side I am referring to is this:
https://github.com/rdkcentral/ThunderClientLibraries/blob/master/Source/ocdm/open_cdm.h#L443 you should be able to call it from within the adapter implementation when you started the server succefully. And it ends up here: https://github.com/rdkcentral/OCDM-Clearkey/blob/master/MediaSession.h#L47 and here in code https://github.com/rdkcentral/OCDM-Clearkey/blob/master/MediaSession.cpp#L138. It intention is indeed to prepare after the creation of the session to "load" anything reuired to make the session viable for handling. Guess it would be a good point to also hookup to the server instance than.
Let us now if this works for you as well...